test/distributed/test_distributed.py

```diff
@@ -2684,6 +2694,427 @@ def test_nccl_gather_object_err(self):
+
+        with net.join(divide_by_initial_world_size=False):
+            for _ in range(n_iters):
+                loss = net(inp).sum()
+                loss.backward()
+                # The grad is always expected_grad, since we divide by the number
+                # of currently active processes and inactive processes contribute
+                # zero gradient. If we kept dividing by static initial world
+                # size as processes leave, the grad would be smaller.
+                expected_grad = torch.ones(dim, dim, device=self.rank) * grad_scale
+                param = list(net.parameters())[0]
+                self.assertEqual(expected_grad, param.grad)
+                # Avoid accumulating grads so that it's the same every iteration
+                net.zero_grad()
+                torch.cuda.synchronize(device=self.rank)
+
+        # If divide_by_initial_world_size=True (default), we always scale grads
+        # by the initial world_size.
+        with net.join(divide_by_initial_world_size=True):
+            for i in range(n_iters):
+                loss = net(inp).sum()
+                loss.backward()
+                effective_ws = dist.get_world_size()
+                if i >= 3:
+                    effective_ws -= 1
+                expected_grad = (
+                    torch.ones(dim, dim, device=self.rank) * grad_scale * effective_ws
+                ) / dist.get_world_size()
+                param = list(net.parameters())[0]
+                self.assertEqual(expected_grad, param.grad)
+                # Avoid accumulating grad so that it's the same every iteration.
+                net.zero_grad()
+                torch.cuda.synchronize(device=self.rank)
+
+    @require_backend({"gloo", "nccl"})
+    @require_backends_available({"gloo", "nccl"})
+    @skip_if_lt_x_gpu(2)
+    @skip_if_rocm
+    def test_ddp_join_model_equivalence(self):
+        # Verifies equivalence with model training locally and with DDP under
+        # the join context manager.
+        batch = 3
+        dim = 10
+        learning_rate = 0.03
+        model = nn.Linear(dim, dim, bias=False)
+        inp = torch.rand(batch, dim, device=self.rank)
+        local_model = copy.deepcopy(model)
+        local_model = local_model.cuda(self.rank)
+        rank_to_iter_mapping = {rank : 2 * (rank + 1) for rank in range(dist.get_world_size())}
+        # run local model
+        local_iters = sum(rank_to_iter_mapping.values())
+        local_optim = torch.optim.SGD(local_model.parameters(), lr=learning_rate)
+        for _ in range(local_iters):
+            local_optim.zero_grad()
+            out = local_model(inp)
+            loss = out.sum()
+            loss.backward()
+            local_optim.step()
+
+        # run DDP model with join API
+        num_iters = rank_to_iter_mapping[self.rank]
+        net = torch.nn.parallel.DistributedDataParallel(
+            model.cuda(self.rank), device_ids=[self.rank]
+        )
```

Comment:
In this version, to maintain equivalence we scale by the static `world_size` after backwards()/allreduce (this is done via the higher learning rate on 2822). In the previous version, this test worked because we would divide by the effective world size in the allreduce logic, but in this test I had some logic to scale by the effective world size (I'm not sure how to view it on GH but it can be viewed by going to V58 on the internal diff). So, the grads in that test and this modified test should be the same.